Automattic/jetpack
# Sync Concurrency / Race Conditions

Turns out the concurrency issue is more complicated than a simple lock. This update overhauls several layers of the Full Sync process to ensure we get a consistent single full sync request sent to WP.com for processing. It handles xmlrpc errors, timeouts and other errors gracefully.

With this implementation it is still possible for the site to send concurrent requests to WP.com for processing. However when this happens WP.com will respond with a Retry-After header to notify the site to delay sending of further requests. The site will then pause sending of data for the specified length of time. 

The Retry-After implementation is queue aware and will allow Incremental Sync actions if Full Sync actions are blocked for concurrency.

1) Separate expiration to clear the lock but not send to reduce race conditions between delete/insert operation.
2) sending of full sync now returns error state so the lock is not cleared if there was a timeout or other xmlrpc issue.
3) Jetpack_IXR_Client updated to read response headers.
4) Actions adds a lock if Retry-After header is present on sync responses.
5) If Retry-After lock present don't send sync actions for queue. Also adds logic to clear lock if expired
6) Update Sync lock option so it is not autoloaded.

Fixes # https://wp.me/p9dueE-2kk

#### Changes proposed in this Pull Request:
* Reduce concurrent sending of data during a Full Sync in high request environments.
* Retry-After adherence in Sync.

#### Does this pull request change what data or activity we track or use?
NO

#### Testing instructions:

* Apply Patch
* Sandbox your Jetpack Connection to your WP.com Sandbox
* Apply D55878-code to your WP.com Sandbox
* on a WP.com sandbox run wp jetpack-sync pull --blog-id=XXXXX --start
* Verify the Full Sync starts and completes.

* Create a test site with 50,000 post and term relationships for every post.
* Trigger a Full Sync (debugger)
* Stress test the site by calling sync/checkout with 10+ different processes.
* Verify when concurrency occurs that the site stop sending data to WP.com for the Retry-After period. 
* Verify Site experiences no degradation in performance.

#### Proposed changelog entry for your changes:
* Sync Performance : Reduce concurrency of sync requests by implementation of Retry-After for concurrent requests.
